### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by klauspdoll**

### CODE REVIEW

**Review by pjknowles — APPROVED** [review]
Yes, I agree. We might want to eventually cache `whole_of_procedures_registry` to avoid potentially reading the registry procedures from file on every editor key stroke. You can also delete the obsolete commented statements - it's actually easier to read the diff if they are not still there.

**Comment by klauspdoll** [comment]
I have done some more changes and would merge it now.

Perhaps you can check if

`self.whole_of_procedures_registry = self.project.procedures_registry()`

is in the right place or not.

File: ProjectWindow.py [hunk 1]
```diff
@@ -102,6 +102,7 @@ def __init__(self, filename, window_manager, latency=1000):
 
         self.webengine_profiles = []
 
+        self.whole_of_procedures_registry = self.project.procedures_registry()
```
Comment (pjknowles) [inline ProjectWindow.py:105]: It's OK, but if you want to make the code even safer, and with related code fragments in the same place, you can do 
```
if self.whole_of_procedures_registry is None:
   self.whole_of_procedures_registry=...
```
and move this code inside `allowed_methods()` where it is needed.

**Review by pjknowles — COMMENTED** [review]
It needs also the following patch to make the checking of precursor methods work properly:
```
% git diff                                                                                                                                                                                                                       ✭
diff --git a/molpro_input.py b/molpro_input.py
index 76a465b..c5b5eb8 100644
--- a/molpro_input.py
+++ b/molpro_input.py
@@ -79,7 +79,7 @@ def parse(input: str, allowed_methods: list, debug=False):
                 # print('field, key=', key, 'value=', value)
                 variables[key] = value.replace('!', ',')  # unprotect
         elif any(
-                [re.match('{? *' + df_prefix + spin_prefix + precursor_method + '}?', command, flags=re.IGNORECASE) for
+                [re.match('{? *' + df_prefix + spin_prefix + precursor_method + '[;}]', command+';', flags=re.IGNORECASE) for
                  df_prefix
                  in
                  df_prefixes
```

**Comment by klauspdoll** [comment]
> It needs also the following patch to make the checking of precursor methods work properly:
> 
> ```
> % git diff                                                                                                                                                                                                                       
> -                [re.match('{? *' + df_prefix + spin_prefix + precursor_method + '}?', command, flags=re.IGNORECASE) for
> +                [re.match('{? *' + df_prefix + spin_prefix + precursor_method + '[;}]', command+';', flags=re.IGNORECASE) for

should it be

 '[;}]'   or  '[;]}'

**Comment by pjknowles** [comment]
> > It needs also the following patch to make the checking of precursor methods work properly:
> > ```
> > % git diff                                                                                                                                                                                                                       
> > -                [re.match('{? *' + df_prefix + spin_prefix + precursor_method + '}?', command, flags=re.IGNORECASE) for
> > +                [re.match('{? *' + df_prefix + spin_prefix + precursor_method + '[;}]', command+';', flags=re.IGNORECASE) for
> > ```
> 
> should it be
> 
> '[;}]' or '[;]}'

Definitely not `'[;]}'` - that would require the `}` to be present. Possibly by the the '}' is in any case stripped. Is there a case where what I wrote goes wrong?

**Comment by klauspdoll** [comment]
> > > It needs also the following patch to make the checking of precursor methods work properly:
> > > ```
> > > % git diff                                                                                                                                                                                                                       
> > > -                [re.match('{? *' + df_prefix + spin_prefix + precursor_method + '}?', command, flags=re.IGNORECASE) for
> > > +                [re.match('{? *' + df_prefix + spin_prefix + precursor_method + '[;}]', command+';', flags=re.IGNORECASE) for
> > > ```
> > 
> > 
> > should it be
> > '[;}]' or '[;]}'
> 
> Definitely not `'[;]}'` - that would require the `}` to be present. Possibly by the the '}' is in any case stripped. Is there a case where what I wrote goes wrong?

emacs does not like it. It wants the left { and the right } to match.

**Comment by pjknowles** [comment]
> emacs does not like it. It wants the left { and the right } to match.

Why does emacs' view of what the contents of `command` might be have any relevance? As far as I understand, in a python regex, if `}` appears inside `[]` it is interpreted just as its own character, not as a group delimiter. You can try for yourself at https://pythex.org . The `}` definitely needs to be accommodated, since `command` will contain it if the input line is, eg, `{rhf}`.

**Comment by klauspdoll** [comment]
I have done the changes you suggested and would merge it now.

**Pull request merged**